### PR TITLE
[experimental] Add decoders, distributions, encoders, layers, networks, and utils

### PR DESCRIPTION
**Pull request by ervteng**
### Proposed change(s)

Add PyTorch core files (decoders.py, distributions.py, encoders.py, layers.py, networks.py, and utils.py) in `torch` folder. Also adds tests for these files, and changes to `test_requirements.txt` required to run them. Note that these files are disconnected to any code, and we plan to slowly merge in files from `develop-add-fire` until functionality is complete. 

### Types of change(s)

- [ ] Bug fix
- [ ] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [x] Other (Core files needed for Torch)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -70,7 +70,7 @@ jobs:
             . venv/bin/activate
             mkdir test-reports
             pip freeze > test-reports/pip_versions.txt
-            pytest -n 2 --cov=ml-agents --cov=ml-agents-envs --cov=gym-unity --cov-report html --junitxml=test-reports/junit.xml -p no:warnings
+            pytest --cov=ml-agents --cov=ml-agents-envs --cov=gym-unity --cov-report html --junitxml=test-reports/junit.xml -p no:warnings
```
Comment (ervteng) [inline .circleci/config.yml:73]: `pytest-xdist` and PyTorch causes a multiprocessing error to be thrown. Didn't notice a huge increase in testing time though didn't look too deeply into it.

File: test_requirements.txt [hunk 2]
```diff
@@ -3,5 +3,8 @@ pytest>4.0.0,<6.0.0
 pytest-cov==2.6.1
 pytest-xdist
 
+# PyTorch tests are here for the time being, before they are used in the codebase.
+torch>=1.5.0
```
Comment (ervteng) [inline test_requirements.txt:7]: Requirement only for CI at the moment.

Comment (chriselion) [reply test_requirements.txt:7]: Came here to check this :)

**Comment by ervteng** [comment]
> I approve of this PR, this does mean that we should not modify these files on `develop-add-fire` anymore or will we be merging updated versions as they change?

It seems really unlikely that this will be true, but hopefully the changes are smaller than this fat one

**Pull request merged**